### PR TITLE
[Misc][LoRA] Clean up the function interface of Punica

### PR DESCRIPTION
**Pull request by jeejeelee**
## Description
The current Punica interfaces(functions whose names start with `add_` in [punica.py](https://github.com/vllm-project/vllm/blob/v0.6.4.post1/vllm/lora/punica.py)) are somewhat complex and disorganized. This PR primarily categorizes all LoRA layers into the following three types and  clean up related interfaces. The benefits  include simplifying and standardizing functions to facilitate LoRA support across other platforms, such as: https://github.com/vllm-project/vllm/pull/4830 https://github.com/vllm-project/vllm/pull/10565
### 1. Linear
 `Three` functions are provided, where [add_lora_linear](https://github.com/jeejeelee/vllm/blob/optimize-punica-function/vllm/lora/punica.py#L629) is for all linear lora in [layers.py](https://github.com/vllm-project/vllm/blob/v0.6.4.post1/vllm/lora/layers.py), while [add_shrink](https://github.com/jeejeelee/vllm/blob/optimize-punica-function/vllm/lora/punica.py#L525) and [add_expand](https://github.com/jeejeelee/vllm/blob/optimize-punica-function/vllm/lora/punica.py#L556)  are designed for full sharded lora in [fully_sharded_layers.py](https://github.com/vllm-project/vllm/blob/v0.6.4.post1/vllm/lora/fully_sharded_layers.py)
### 2. Embedding
Only `one` function is needed, see: [add_lora_embedding](https://github.com/jeejeelee/vllm/blob/optimize-punica-function/vllm/lora/punica.py#L603)
### 3.  Logits
Only `one` function is needed,see:[add_lora_logits](https://github.com/jeejeelee/vllm/blob/optimize-punica-function/vllm/lora/punica.py#L685)

cc @SanjuCSudhakaran  
ping @DarkLight1337 
@Isotr0py Since you're the author of #4830, would you please take a look at this PR as well?

### CODE REVIEW

File: vllm/lora/punica.py [hunk 1]
```diff
@@ -450,41 +450,35 @@ def expand_slice_decode(
...
-        output: torch.Tensor,
-        bias_stacked: torch.Tensor,
-    ):
-        """Applies bias to output
-
-        Input shapes:
-            bias_stacked:    (num_loras, output_dim)
-            indices:         (batch_size)
-            output:          (batch_size, output_dim)
+    def apply_expand(self,
```
Comment (Isotr0py) [inline vllm/lora/punica.py:453]:
```suggestion
    def _apply_expand(self,
```

TBH, I sometimes feel confused about the methods like `apply_bias` etc, because it is just an internal method used in `PunicaWrapper`, but it quite looks like a method that we will call with `apply` in `layers.py`.

Perhaps renaming it to `_apply` would be clearer?

**Review by Isotr0py — COMMENTED** [review]
Thanks for cleaning up the duplicate function! I added some initial comments about the punica interface. PTAL!

File: vllm/lora/punica.py (hunk 1 above)
Comment (jeejeelee) [reply vllm/lora/punica.py:453]: Make sense. 
BTW, I have a  prototype that modifies the punica wrapper, you can take a look at it, see: https://github.com/jeejeelee/vllm/tree/refactor-punica/vllm/lora

File: vllm/lora/layers.py [hunk 2]
```diff
@@ -568,16 +539,22 @@ class MergedColumnParallelLinearWithLoRA(ColumnParallelLinearWithLoRA):
 
     def __init__(self, base_layer: MergedColumnParallelLinear) -> None:
         super().__init__(base_layer)
+        # There are two LoRA layers
+        self.n_slices = 2
```
[the remark this answers is not in the document]
Comment (jeejeelee) [reply vllm/lora/layers.py]: Done

**Review by Isotr0py — APPROVED** [review]
Overall LGTM!
